### PR TITLE
Use newer resolve API from coursier

### PR DESCRIPTION
**Pull request by lefou**
As part of experimenting the solution space for feature request https://github.com/com-lihaoyi/mill/issues/3075, I found that some resolution validation errors are only correctly detected, when the newer coursier API is used. Therefore I extracted some commits of PR https://github.com/com-lihaoyi/mill/pull/3084 into this standalone PR and fixed any issues on the way.

Changes:
* Use newer coursier `Resolve` API to really detect conflicting dependencies
* Introduce a new `CoursierModule.enforceSameDependencyVersions`, which allows to sync versions for artifacts belonging to the same release, e.g. `scala-library` and `scala-reflect`.

### CODE REVIEW

File: main/util/src/mill/util/CoursierSupport.scala [hunk 1]
```diff
@@ -208,60 +227,126 @@ trait CoursierSupport {
...
+        ctx = ctx,
+        errorMsgExtractor = (r: Resolution) => r.errors.flatMap(_._2)
+      )(resolve.run)
 
-    val resolution =
-      retry(ctx = ctx, errorMsgExtractor = (r: Resolution) => r.errors.flatMap(_._2)) {
-        () => start.process.run(fetch).unsafeRun()
-      }
+      val resolution = resolution0
+        .withMapDependencies(mapDependencies)
```
Comment (lefou) [inline main/util/src/mill/util/CoursierSupport.scala:314]: I think this `mapDependencies` is applied too late. It must be applied before resolution instead. But I can't find any other way in the new `Resolve` API to pass a map function. @alexarchambault Do you have an idea?

File: main/util/src/mill/util/CoursierSupport.scala [hunk 2]
```diff
@@ -208,60 +227,126 @@ trait CoursierSupport {
...
+      .map(d => mapDependencies.fold(d)(_.apply(d)))
       .map { d => d.module -> d.version }
       .toMap
 
-    val start0 = Resolution()
-      .withRootDependencies(
-        deps.iterator.map(mapDependencies.getOrElse(identity[Dependency](_))).toSeq
-      )
-      .withForceVersions(forceVersions)
-      .withMapDependencies(mapDependencies)
```
Comment (lefou) [inline main/util/src/mill/util/CoursierSupport.scala:235]: For this `withMapDependencies` I couldn't find an equivalent in the new API.

**Comment by alexarchambault** [comment]
FYI I'm giving a try at this too, I should open a PR for it soon

**Comment by lefou** [comment]
* https://github.com/com-lihaoyi/mill/pull/3513

File: main/util/src/mill/util/CoursierSupport.scala (hunk 2 above)
Comment (lefou) [reply main/util/src/mill/util/CoursierSupport.scala:235]: Related?

* https://github.com/coursier/coursier/pull/3072

File: main/util/src/mill/util/CoursierSupport.scala (hunk 1 above)
Comment (lefou) [reply main/util/src/mill/util/CoursierSupport.scala:314]: Related?

* https://github.com/coursier/coursier/pull/3072

File: main/util/src/mill/util/CoursierSupport.scala (hunk 2 above)
Comment (alexarchambault) [reply main/util/src/mill/util/CoursierSupport.scala:235]: I hadn't realized how deeply ingrained in Mill `mapDependencies` was (it's basically exposed to Mill users). On the [sbt side](https://github.com/coursier/sbt-coursier), it's only used internally, so these uses were converted to more explicit parameters in [`ResolutionParams`](https://github.com/coursier/coursier/blob/36341ddc7701bd38c3f8ac0e0e89ac116a2e2921/modules/coursier/shared/src/main/scala/coursier/params/ResolutionParams.scala) (`typelevel`, `doForceScalaVersion`, `doOverrideFullSuffix`, `selectedScalaVersion`).

https://github.com/coursier/coursier/pull/3072 adds back the possibility for API users to transform dependencies this way, yes.

**Comment by lihaoyi** [comment]
Closing in favor of https://github.com/com-lihaoyi/mill/pull/3513

**Comment by lefou** [comment]
We probably want to add the `enforceSameDependencyVersions` target to `ScalaModule` in a separate PR.

**Pull request closed without merging**